sound/soc/sof/intel/hda-ipc.c

```diff
@@ -172,16 +172,14 @@ irqreturn_t hda_dsp_ipc_irq_thread(int irq, void *context)
 					HDA_DSP_REG_HIPCCTL,
 					HDA_DSP_REG_HIPCCTL_DONE, 0);
 
-		/* handle immediate reply from DSP core - ignore ROM messages */
-		if (hda_dsp_ipc_is_sof(msg)) {
-			hda_dsp_ipc_get_reply(sdev);
-			snd_sof_ipc_reply(sdev, msg);
-		}
-
-		/* wake up sleeper if we are loading code */
+		/* wake up sleeper and no reply if we are loading code */
 		if (sdev->code_loading)	{
 			sdev->code_loading = 0;
 			wake_up(&sdev->waitq);
+		} else {
+			/* handle immediate reply from DSP core */
+			hda_dsp_ipc_get_reply(sdev);
+			snd_sof_ipc_reply(sdev, msg);
```

Comment:
@plbossart @xun2z I feel that hda_dsp_ipc_get_reply() ignores the reply message that arrives when not expected and  snd_sof_ipc_reply() should do the same too.

@xun2z seems reasonable with the assumption that the ROM messages will only show up during code loading. Im not sure about this though.

Reply:
@xun2z The assumption is there is no ipc before firmware loading finish. I think it is right. But it seems also assume that the firmware loading finished when the first irq after starting loading firmware arrived. Is there any corner case that firmware loading is not finished when the first irq after starting loading firmware arrived? And I think we still need the ``hda_dsp_ipc_is_sof(msg)`` check even the flow is changed.

Reply:
@bardliao This is true, I assumed 2 things. But my second assumption is bolder: 

1. We have no ipc before firmware loading finish
2. It is the first irq tell us firmware load is down

This is the first reason that I take out `hda_dsp_ipc_is_sof(msg)`. But now I think keep it there will be a better choice.

Reply:
why not apply the change in hda-ipc.c to cnl.c
````
	+	if (hda_dsp_ipc_is_sof(msg)) {
			hda_dsp_ipc_get_reply(sdev);
			snd_sof_ipc_reply(sdev, msg);
	+	}
````

Reply:
please check https://github.com/thesofproject/linux/commit/abfab3266c93f33563ccb76b308198b81b3aa102#diff-0c41ff728d2838606e1e70839d291aed

````
The currently ignored IPC 0x1004000 is only one of several possible
ROM IPC messages. Detect and ignore all of them.
````
I think @lyakh assumed hda-ipc.c would cover apollolake, cannonlake  and later chips, so he only refined the hda-ipc.c and missed cnl.c

Reply:
Mostly, I think the patch should be OK for our current assumption.
But I still have the below concern:
All the platforms definitely will raise an interrupt after loading FW?
If you look into the fw loading code in hda-loader.c, it can handle the situation that the ROM doesn't send any interrupt after fw loading or FW trigger the interrupt later than the timeout.
If the FW doesn't raise such interrupt, you patch will block sof driver working.

Reply:
@RanderWang you're right, thanks for spotting, I missed the CNL. @xun2z if possible, I think it would be better to use the same approach on CNL as is already used on other HDA platforms.